src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +72,32 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+static char *utoa_upper(unsigned value, char *str, int base)
+{
+	const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+	/* Check base is supported. */
+	if ((base < 2) || (base > 36)) {
+		str[0] = '\0';
+		return NULL;
+	}
+
+	/* Convert to string. Digits are in reverse order.  */
+	int l = 0;
+	do {
+		const unsigned remainder = value % base;
+		str[l++] = digits[remainder];
+		value = value / base;
+	} while (value != 0);
+	str[l] = '\0';
+
+	/* Reverse string.  */
+	for (int j = 0, i = l - 1; j < i; j++, i--) {
```

Comment:
Please give this loop counter a more descriptive name, not `i` and `j`.

Right now we cannot easily tell what this is doing and why it might be correct, and because it's using `int` this is subject to overflow UB if something goes wrong.

Reply:
This loop runs for half the string length iterations and swaps chars in palindrome fashion (in-place, no allocations) -- but with your other suggestion it can be dropped in favor of actually using more memory (for the static writable buffer).

Reply:
Rewritten the reversing part to read simpler, and now it has to do a full copy between different buffers instead of a half-copy in-place. No big deal, it's a short string in DFU descriptors context.